File: crates/ferrogate-gateway/src/server/rbac_test.rs
// Token4AI Cloud Attribution
// Developed by the commercial cloud service company represented by https://token4ai.cloud.
// Author: jamesduan (X: https://x.com/JamesDuanL)
// Created: 2026-07-27
// description: Issue #543 -- the storage-FAILURE branch of `rbac_catalog_scope`,
// the third property of #518 and the only one nothing held. The five #518
// tests in ferrogate-cli/tests/rbac_catalog_scope_admin_api.rs all stay green
// while the resolver degrades a failed control-plane read to
// `RbacCatalogScope::Full`, which hands every tenant-scoped `admin.read` key
// the entire platform RBAC catalog -- the exact disclosure #518 closed,
// reached through the error path instead of the happy path.

use super::*;
use crate::tenant_scope_reads::fault::{
    block_on, platform_operator_auth, tenant_auth, FaultyTenantScopeReads, TenantScopeRead,
};

/// A store that would resolve to a genuinely NARROW tenant scope if every read
/// succeeded: tenant-a holds `role-a` only, and the catalog also carries
/// `role-b` (bound to nobody here) with a permission tenant-a must never see.
///
/// Every failure test below starts from this store and arms one read, so a
/// resolver that degrades to `Full` is distinguishable from one that resolves
/// correctly -- the two answers differ on `role-b`/`billing.admin`.
fn catalog() -> FaultyTenantScopeReads {
    FaultyTenantScopeReads::healthy()
        .with_role("role-a", &["chat.read"])
        .with_role("role-b", &["billing.admin"])
        .with_binding("tenant-a", "role-a")
        .with_binding("tenant-b", "role-b")
}

/// Baseline, and the guard that makes every failure case below non-vacuous:
/// with healthy storage this resolver returns a NARROW tenant scope, so a
/// later `assert!(is_err())` is testing the failure branch rather than a store
/// that cannot answer anything.
///
/// Pins: the `role_ids.contains(&role.id)` filter and the
/// `CallerScope::Tenant` branch of `rbac_catalog_scope`.
/// Catches: dropping the role filter (tenant-a would gain `billing.admin`),
/// and returning `Full` for a tenant-scoped caller (`is_full()` would flip).
#[test]
fn healthy_storage_resolves_a_narrow_tenant_scope() {
    let store = catalog();
    let scope = block_on(rbac_catalog_scope(&store, &tenant_auth("tenant-a")))
        .expect("healthy storage must resolve a scope");

    assert!(
        !scope.is_full(),
        "a tenant-scoped caller must never resolve to the unfiltered catalog"
    );
    match &scope {
        RbacCatalogScope::Tenant {
            role_ids,
            permission_keys,
        } => {
            assert!(
                role_ids.contains("role-a"),
                "own bound role must be visible"
            );
            assert!(
                !role_ids.contains("role-b"),
                "another tenant's role must not be visible: {role_ids:?}"
            );
            assert!(permission_keys.contains("chat.read"));
            assert!(
                !permission_keys.contains("billing.admin"),
                "permission keys must be limited to the caller's own roles: \
                 {permission_keys:?}"
            );
        }
        RbacCatalogScope::Full => unreachable!("checked by is_full above"),
    }
    assert_eq!(
        store.reads(),
        vec![
            TenantScopeRead::ListTenantRoleBindings,
            TenantScopeRead::ListRoles
        ],
        "the resolver must read the bindings and the role catalog, in that order"
    );
}

/// THE #543 property, read one: a failed `list_tenant_role_bindings` must
/// propagate.
///
/// Pins: `state.list_tenant_role_bindings(tenant_id).await?` in
/// `rbac_catalog_scope`.
/// Catches: the mutation the test gate demonstrated on #518 --
/// `let Ok(bindings) = state.list_tenant_role_bindings(tenant_id).await else {
/// return Ok(RbacCatalogScope::Full) };` -- and equally
/// `.unwrap_or_default()`, which degrades to an empty (silently
/// everything-denied) scope. Both turn the `Err` this asserts into an `Ok`.
#[test]
fn failed_binding_read_propagates_and_never_degrades_to_a_scope() {
    let store = catalog().failing(TenantScopeRead::ListTenantRoleBindings);

    let resolved = block_on(rbac_catalog_scope(&store, &tenant_auth("tenant-a")));

    match resolved {
        Err(_) => {}
        Ok(RbacCatalogScope::Full) => panic!(
            "a failed control-plane read degraded to the UNFILTERED catalog: every \
             tenant-scoped admin.read key would receive the whole platform RBAC \
             catalog through the error path (#518 reopened via #543)"
        ),
        Ok(RbacCatalogScope::Tenant { role_ids, .. }) => panic!(
            "a failed control-plane read degraded to a fabricated tenant scope \
             ({role_ids:?}); an unavailable control plane is not an answer about \
             what this tenant may see"
        ),
    }
    assert!(
        store
            .reads()
            .contains(&TenantScopeRead::ListTenantRoleBindings),
        "the failure must come from the binding read actually being attempted"
    );
}

/// THE #543 property, read two: the SECOND storage read has the same
/// obligation as the first. It is a separate `?` on a separate line, so it is
/// a separate assertion.
///
/// Pins: `state.list_roles().await?` in `rbac_catalog_scope`.
/// Catches: `.unwrap_or_default()` on the role list, which resolves the caller
/// to a scope holding its role ids but NO permission keys -- and, worse, any
/// `else { return Ok(RbacCatalogScope::Full) }` on this read.
#[test]
fn failed_role_catalog_read_propagates_and_never_degrades_to_a_scope() {
    let store = catalog().failing(TenantScopeRead::ListRoles);

    let resolved = block_on(rbac_catalog_scope(&store, &tenant_auth("tenant-a")));

    match resolved {
        Err(_) => {}
        Ok(RbacCatalogScope::Full) => {
            panic!("a failed role-catalog read degraded to the UNFILTERED catalog")
        }
        Ok(RbacCatalogScope::Tenant {
            permission_keys, ..
        }) => panic!(
            "a failed role-catalog read degraded to a scope with permission keys \
             {permission_keys:?} rather than refusing to answer"
        ),
    }
    assert!(
        store.reads().contains(&TenantScopeRead::ListRoles),
        "the failure must come from the role-catalog read actually being attempted"
    );
}

/// The other half of "propagate": propagation is only safe if the CALLER turns
/// the refusal into a refusal. All four `/admin/v1/permissions*` /
/// `/admin/v1/roles*` GETs map a failed `rbac_catalog_scope` through this one
/// function.
///
/// Pins: `rbac_scope_storage_error`, and through it the four `Err(error) =>`
/// arms that now call it.
/// Catches: any remapping of the failure onto a success-shaped or permissive
/// response (a 200 with an empty list, or a 403), and any drift of the error
/// code the admin console keys off.
#[test]
fn a_failed_scope_becomes_503_storage_unavailable() {
    let store = catalog().failing(TenantScopeRead::ListTenantRoleBindings);
    let error = block_on(rbac_catalog_scope(&store, &tenant_auth("tenant-a")))
        .err()
        .expect("a failed control-plane read must not resolve to a scope");

    let (status, code, message) = rbac_scope_storage_error(&error);

    assert_eq!(status, StatusCode::SERVICE_UNAVAILABLE);
    assert_eq!(code, "storage_unavailable");
    assert!(
        !message.is_empty(),
        "the refusal must carry the underlying storage error"
    );
}

/// The platform-operator short-circuit is NOT collateral damage of the
/// fail-closed rule: a declared platform operator resolves to `Full` without
/// reading storage at all, so a control-plane blip cannot 503 the operator
/// surface that is used to diagnose it.
///
/// Pins: `let CallerScope::Tenant(tenant_id) = auth.caller_scope() else {
/// return Ok(RbacCatalogScope::Full) };`.
/// Catches: routing platform operators through the tenant path (the reads are
/// armed to fail, so the resolver would return `Err` instead of `Full`), and
/// any added storage read before the classification (the `reads()` assertion).
#[test]
fn platform_operator_resolves_full_without_touching_storage() {
    let store = catalog()
        .failing(TenantScopeRead::ListTenantRoleBindings)
        .failing(TenantScopeRead::ListRoles);

    let scope = block_on(rbac_catalog_scope(&store, &platform_operator_auth()))
        .expect("a declared platform operator must not depend on a storage read");

    assert!(scope.is_full());
    assert!(
        store.reads().is_empty(),
        "the platform-operator answer must be decided from the credential alone, \
         not from storage: {:?}",
        store.reads()
    );
}

File: apps/gateway/test/rbac.test.ts
/**
 * `D1RbacAuthorizer` — the durable Permission → Role → TenantRoleBinding walk.
 *
 * The port of `crates/ferrogate-gateway/src/state_rbac.rs::
 * tenant_has_permission_result`, run against the REAL `CONTROL_DB` binding
 * `wrangler.toml` declares and the REAL `permissions` / `roles` /
 * `tenant_role_bindings` DDL from `sql/d1-ts/control/0001_init_control.sql`
 * (applied by `test/setup-d1.ts`). Nothing here doubles the database — the
 * outage tests substitute a THROWING database, which is the one thing a live
 * binding cannot be asked to do.
 *
 * The file has two halves and they prove different things:
 *
 *  - the decision table, driven directly against the adapter;
 *  - the MOUNT, driven through `SELF.fetch` — i.e. through `depsFromEnv` on the
 *    app `src/index.ts` exports. A `D1RbacAuthorizer` that is written, tested
 *    and never reached by `depsFromEnv` is the defect class this repo has
 *    already shipped twice, so the mount gets its own failing-when-unmounted
 *    assertion rather than being assumed.
 */
import { SELF, env } from "cloudflare:test";
import { afterEach, beforeAll, beforeEach, describe, expect, it } from "vitest";

import {
  ConfiguredRbacAuthorizer,
  D1RbacAuthorizer,
  RBAC_PERMISSION_EXISTS_SQL,
  type RbacDatabase,
  depsFromEnv,
} from "../src/adapters.js";
import { CONTROL_DATA_ADDRESS } from "@ferrogate/storage";
import { CONTROL_STORAGE_MISCONFIGURED, controlDatabaseFrom } from "../src/control-data.js";
import type { AuthContext, RbacDecision } from "../src/ports.js";
import {
  seedTenantRosterRows,
  tenantObjectPrivilegedBatch,
  tenantObjectRouter,
} from "./tenant-object.js";

const bindings = env as unknown as Record<string, unknown>;

function controlDb(): D1Database {
  const binding = bindings.CONTROL_DB as D1Database | undefined;
  if (binding === undefined) {
    // Loud, never a silent skip: `wrangler.toml` declares it, so an absent
    // binding means the declaration was removed and this suite is about to
    // prove something other than what it claims.
    throw new Error(
      "rbac tests expect the `CONTROL_DB` D1 binding (apps/gateway/wrangler.toml). " +
        "See src/adapters.ts for why RBAC reads the CONTROL database.",
    );
  }
  return binding;
}

/** A tenant credential, the shape `contractAuth` puts on the context. */
function tenantAuth(tenantId: string): AuthContext {
  return {
    subject: "key_x",
    tenancy: { tenantId, projectId: null, workspaceId: null, userId: null },
    scopes: ["admin.read"],
    platformOperator: false,
    source: "durable_native",
  };
}

const OPERATOR: AuthContext = {
  subject: "key_root",
  tenancy: { tenantId: null, projectId: null, workspaceId: null, userId: null },
  scopes: ["*"],
  platformOperator: true,
  source: "static_config",
};

/** Seed a permission row (step 1 of the Rust walk: the action must be declared). */
async function declarePermission(key: string): Promise<void> {
  await controlDb()
    .prepare(
      "INSERT OR REPLACE INTO permissions (id, key, name, description) VALUES (?1, ?2, ?3, '')",
    )
    .bind(`perm_${key}`, key, key)
    .run();
}

/** Seed a role holding `permissionKeys`, and bind it to `tenantId`. */
async function grantRole(
  roleId: string,
  permissionKeys: readonly string[],
  tenantId?: string,
): Promise<void> {
  await controlDb()
    .prepare(
      "INSERT OR REPLACE INTO roles (id, name, slug, permission_keys_json) VALUES (?1, ?2, ?3, ?4)",
    )
    .bind(roleId, roleId, roleId, JSON.stringify(permissionKeys))
    .run();
  if (tenantId !== undefined) {
    await bindRole(tenantId, roleId);
  }
}

/**
 * Bind a role id to a tenant — the id may name a role row that does not exist.
 *
 * Post-#863 (`sql/d1-ts/control/0016_tenant_configuration_policy_legacy.sql`)
 * the binding lives in the TENANT OBJECT, not the control database, and the
 * object refuses ordinary tenant SQL against it — so the seed goes through the
 * privileged operator RPC, the same path the production backfill uses.
 */
async function bindRole(tenantId: string, roleId: string): Promise<void> {
  await tenantObjectPrivilegedBatch(tenantId, [
    {
      sql: "INSERT OR REPLACE INTO tenant_role_bindings (id, tenant_id, role_id) VALUES (?, ?, ?)",
      params: [`${tenantId}:${roleId}`, tenantId, roleId],
    },
  ]);
}

/** Every tenant a test in this file writes a durable role binding for. */
const BOUND_TENANTS = ["tenant_a", "tenant_c"] as const;

async function resetRbacTables(): Promise<void> {
  const db = controlDb();
  // The shared catalog stays in CONTROL; the bindings are tenant-object rows.
  await db.batch([db.prepare("DELETE FROM roles"), db.prepare("DELETE FROM permissions")]);
  for (const tenantId of BOUND_TENANTS) {
    await tenantObjectPrivilegedBatch(tenantId, [
      { sql: "DELETE FROM tenant_role_bindings" },
      { sql: "DELETE FROM tenant_role_catalog" },
    ]);
  }
}

/** A database that fails every statement, as an outage would. */
const failingDb: RbacDatabase = {
  prepare(): never {
    throw new Error("D1_ERROR: control database is unreachable");
  },
};

/** A database that records whether it was consulted at all. */
function recordingDb(): { db: RbacDatabase; readonly statements: string[] } {
  const statements: string[] = [];
  const inner = controlDb() as unknown as RbacDatabase;
  return {
    statements,
    db: {
      prepare(sql: string) {
        statements.push(sql);
        return inner.prepare(sql);
      },
    },
  };
}

const ACTION = "guardrails.policy.read";

beforeEach(async () => {
  await resetRbacTables();
});

// ---------------------------------------------------------------------------
// The decision table
// ---------------------------------------------------------------------------

describe("D1RbacAuthorizer — tenant_has_permission_result", () => {
  it("grants when a bound role holds the declared permission", async () => {
    await declarePermission(ACTION);
    await grantRole("role_guardrail_reader", [ACTION], "tenant_a");

    const rbac = new D1RbacAuthorizer(controlDb() as unknown as RbacDatabase, {
      tenantDatabases: tenantObjectRouter(),
    });
    expect(await rbac.authorize(tenantAuth("tenant_a"), ACTION)).toEqual({ allowed: true });
  });

  it("denies a permission that no `permissions` row declares, even when a role names it", async () => {
    // Step 1 of the Rust walk, and it is not a formality: without it a typo in
    // a role's `permission_keys` would mint an entitlement nobody declared.
    await grantRole("role_guardrail_reader", [ACTION], "tenant_a");

    const rbac = new D1RbacAuthorizer(controlDb() as unknown as RbacDatabase, {
      tenantDatabases: tenantObjectRouter(),
    });
    expect(await rbac.authorize(tenantAuth("tenant_a"), ACTION)).toMatchObject({
      allowed: false,
      code: "guardrail_rbac_denied",
    });
  });

  it("denies when the tenant's roles do not hold the permission", async () => {
    await declarePermission(ACTION);
    await grantRole("role_other", ["guardrails.policy.archive"], "tenant_a");

    const rbac = new D1RbacAuthorizer(controlDb() as unknown as RbacDatabase, {
      tenantDatabases: tenantObjectRouter(),
    });
    expect(await rbac.authorize(tenantAuth("tenant_a"), ACTION)).toEqual({
      allowed: false,
      code: "guardrail_rbac_denied",
      message: "tenant roles do not grant required action guardrails.policy.read",
    });
  });

  it("scopes grants to the BOUND tenant — another tenant's role is not a grant", async () => {
    await declarePermission(ACTION);
    await grantRole("role_guardrail_reader", [ACTION], "tenant_a");

    const rbac = new D1RbacAuthorizer(controlDb() as unknown as RbacDatabase, {
      tenantDatabases: tenantObjectRouter(),
    });
    expect(await rbac.authorize(tenantAuth("tenant_b"), ACTION)).toMatchObject({ allowed: false });
  });

  it("skips a binding whose role row is missing rather than treating it as a grant", async () => {
    // Rust: `let Some(role) = get_role(..) else { continue }`.
    await declarePermission(ACTION);
    await bindRole("tenant_a", "role_that_was_deleted");

    const rbac = new D1RbacAuthorizer(controlDb() as unknown as RbacDatabase, {
      tenantDatabases: tenantObjectRouter(),
    });
    expect(await rbac.authorize(tenantAuth("tenant_a"), ACTION)).toMatchObject({ allowed: false });
  });

  it("has NO wildcard: a role holding `*` grants only the permission named `*`", async () => {
    // The Rust compares `key == permission_key`. The var table's `"*"` is a
    // bootstrap convenience and copying it here would be a silent privilege
    // escalation for every role somebody wrote `*` into.
    await declarePermission(ACTION);
    await grantRole("role_star", ["*"], "tenant_a");

    const rbac = new D1RbacAuthorizer(controlDb() as unknown as RbacDatabase, {
      tenantDatabases: tenantObjectRouter(),
    });
    expect(await rbac.authorize(tenantAuth("tenant_a"), ACTION)).toMatchObject({ allowed: false });
  });

  it("names the denial after the action's domain", async () => {
    await declarePermission("workers.self_hosted");

    const rbac = new D1RbacAuthorizer(controlDb() as unknown as RbacDatabase, {
      tenantDatabases: tenantObjectRouter(),
    });
    expect(await rbac.authorize(tenantAuth("tenant_a"), "workers.self_hosted")).toMatchObject({
      allowed: false,
      code: "rbac_denied",
    });
  });

  it("ignores a corrupt role document without revoking the tenant's other roles", async () => {
    await declarePermission(ACTION);
    await controlDb()
      .prepare(
        "INSERT OR REPLACE INTO roles (id, name, slug, permission_keys_json) VALUES (?1, ?1, ?1, ?2)",
      )
      .bind("role_corrupt", "{not json")
      .run();
    await bindRole("tenant_a", "role_corrupt");
    await grantRole("role_good", [ACTION], "tenant_a");

    const rbac = new D1RbacAuthorizer(controlDb() as unknown as RbacDatabase, {
      tenantDatabases: tenantObjectRouter(),
    });
    expect(await rbac.authorize(tenantAuth("tenant_a"), ACTION)).toEqual({ allowed: true });
  });

  it("waves a platform operator through WITHOUT querying the database", async () => {
    const recording = recordingDb();
    const rbac = new D1RbacAuthorizer(recording.db);
    expect(await rbac.authorize(OPERATOR, ACTION)).toEqual({ allowed: true });
    // Not just "allowed": the Rust guards the whole block with
    // `if let CallerScope::Tenant(..)`, so root access must not depend on the
    // control database being up.
    expect(recording.statements).toEqual([]);
  });

  it("issues exactly the two ported statements, in order", async () => {
    // The native-binding compatibility leg (no `tenantDatabases` router): the
    // statement TEXT and ORDER are the ported Rust walk, and both are pinned —
    // `test/fleet-control-matrix.test.ts` scans these very literals to see the
    // gateway reading the RBAC authority. Post-#863 the deployed posture routes
    // the second read through the tenant object instead (the tests above), but
    // the compat leg must keep issuing the permission-existence check FIRST.
    await declarePermission(ACTION);
    const recording = recordingDb();
    await new D1RbacAuthorizer(recording.db).authorize(tenantAuth("tenant_a"), ACTION);
    expect(recording.statements[0]).toBe(RBAC_PERMISSION_EXISTS_SQL);
    expect(recording.statements).toHaveLength(2);
  });
});

// ---------------------------------------------------------------------------
// Outage vs decision, and the config fallback
// ---------------------------------------------------------------------------

describe("D1RbacAuthorizer — an outage is not a decision", () => {
  it("answers `unavailable` (→ 503) when the control database fails", async () => {
    const rbac = new D1RbacAuthorizer(failingDb);
    const decision = await rbac.authorize(tenantAuth("tenant_a"), ACTION);
    expect(decision.allowed).toBe("unavailable");
    expect((decision as { detail: string }).detail).toContain("unreachable");
  });

  it("does NOT fall through to the config table on an outage", async () => {
    // The `D1ApiKeyResolver` rule, and the Rust one: papering an outage over
    // with config would make a control-plane failure grant whatever the
    // bootstrap var happens to say, which is the opposite of fail-closed.
    const generous = new ConfiguredRbacAuthorizer({ tenant_a: ["*"] });
    const rbac = new D1RbacAuthorizer(failingDb, { fallback: generous });
    expect((await rbac.authorize(tenantAuth("tenant_a"), ACTION)).allowed).toBe("unavailable");
  });

  it("DOES fall through to the config table on a durable not-granted", async () => {
    await declarePermission(ACTION);
    const generous = new ConfiguredRbacAuthorizer({ tenant_a: [ACTION] });
    const rbac = new D1RbacAuthorizer(controlDb() as unknown as RbacDatabase, {
      fallback: generous,
      tenantDatabases: tenantObjectRouter(),
    });
    expect(await rbac.authorize(tenantAuth("tenant_a"), ACTION)).toEqual({ allowed: true });
    // …and the fallback is not a blanket yes: a tenant it does not name is
    // still denied.
    expect(await rbac.authorize(tenantAuth("tenant_z"), ACTION)).toMatchObject({ allowed: false });
  });
});

// ---------------------------------------------------------------------------
// The mount
// ---------------------------------------------------------------------------

describe("depsFromEnv wires the durable authorizer", () => {
  it("builds a D1RbacAuthorizer when CONTROL_DATA is bound, and not otherwise", () => {
    expect(depsFromEnv(bindings as never).rbac).toBeInstanceOf(D1RbacAuthorizer);
    // Removing the `[[durable_objects.bindings]] CONTROL_DATA` stanza from
    // wrangler.toml, or the `D1RbacAuthorizer.fromEnv(...) ??` from
    // `depsFromEnv`, turns the first assertion red while leaving the whole
    // decision table above green.
    const withoutBinding = { ...bindings, CONTROL_DATA: undefined };
    expect(depsFromEnv(withoutBinding as never).rbac).toBeInstanceOf(ConfiguredRbacAuthorizer);
  });
});

describe("CONTROL storage seam", () => {
  it("resolves the CONTROL_DATA facade, and rejects d1_compat and other illegal modes", async () => {
    // Zero-D1 S5 (#881): `durable_object` is the ONLY posture. There is no
    // `d1_compat` leg and no `CONTROL_DB`/`BILLING_DB` fallback — `d1_compat` is
    // now an illegal value like any other, not a rollback path.
    let addressedName: string | undefined;
    const namespace = {
      idFromName(name: string) {
        addressedName = name;
        return "control-id";
      },
      get() {
        return {
          query: async () => ({ results: [] }),
          batch: async () => [],
        };
      },
    };

    const facade = controlDatabaseFrom({ CONTROL_DATA: namespace });
    expect(facade).toBeDefined();
    // The facade addresses the singleton "control" object when first used.
    await (facade as D1Database).prepare("SELECT 1").all();
    expect(addressedName).toBe(CONTROL_DATA_ADDRESS);

    // No CONTROL_DATA bound ⇒ undefined (the optional reads' config path),
    // never a fall-through to a legacy binding.
    expect(controlDatabaseFrom({})).toBeUndefined();

    for (const mode of ["d1_compat", "invalid"]) {
      let error: unknown;
      try {
        controlDatabaseFrom({ GATEWAY_CONTROL_STORAGE: mode, CONTROL_DATA: namespace });
      } catch (caught) {
        error = caught;
      }
      expect(error, `mode ${mode} must be refused`).toMatchObject({
        status: 503,
        code: CONTROL_STORAGE_MISCONFIGURED,
      });
    }
  });
});

/**
 * The behavioural half — through `SELF.fetch`, i.e. the deployed app.
 *
 * `/admin/v1/guardrail-policies` is owned by `apps/control-plane`, so this
 * Worker mounts no route for it. That is exactly what makes it a clean probe:
 * `contractAuth` is an `app.use("*")` guard, so it runs and decides BEFORE the
 * 404. A caller that clears the guard sees 404 (nothing here serves it); a
 * caller the RBAC walk denies sees 403 with the Rust's own code. The two are
 * unambiguous.
 *
 * `tenant_c` is used deliberately: `vitest.config.ts` grants
 * `guardrails.policy.read` to `tenant_a` through `TENANT_RBAC_ACTIONS`, so a
 * test on `tenant_a` could pass on the config fallback alone and prove nothing
 * about the durable leg.
 */
describe("composition root — the durable RBAC walk decides a real request", () => {
  const ADMIN_KEY = "fg_tenant_c_admin";

  beforeAll(async () => {
    // A native key for `tenant_c` carrying the operation's `admin.read` scope,
    // so the request reaches the RBAC step instead of stopping at `scope_denied`.
    bindings.GATEWAY_NATIVE_API_KEYS = JSON.stringify([
      { key: ADMIN_KEY, id: "key_tenant_c", tenant_id: "tenant_c", scopes: ["admin.read"] },
    ]);
    // `tenant_c` is not a vitest.config fixture tenant, so `test/setup-d1.ts`
    // seeds no roster row for it — and an ADMITTED request then 503s downstream
    // when the backend-dispatching router cannot place the tenant.
    await seedTenantRosterRows(["tenant_c"]);
  });

  afterEach(async () => {
    await resetRbacTables();
  });

  function listPolicies(): Promise<Response> {
    return SELF.fetch("https://gw.test/admin/v1/guardrail-policies", {
      headers: { authorization: `Bearer ${ADMIN_KEY}` },
    });
  }

  it("403 `guardrail_rbac_denied` with no durable grant and no config grant", async () => {
    const res = await listPolicies();
    expect(res.status).toBe(403);
    expect(((await res.json()) as { error: { code: string } }).error.code).toBe(
      "guardrail_rbac_denied",
    );
  });

  it("clears the guard once a role in the CONTROL database grants the action", async () => {
    await declarePermission(ACTION);
    await grantRole("role_guardrail_reader", [ACTION], "tenant_c");

    const res = await listPolicies();
    // 404, not 403: the guard admitted the request and `apps/control-plane`
    // owns the route. Nothing but the D1 rows above changed between the two
    // tests, so this is the durable walk deciding, not the var table.
    expect(res.status).toBe(404);
  });

  it("503 `rbac_unavailable` is what an unmigrated control database answers", async () => {
    // Proven by dropping the table the walk reads first, then restoring it —
    // the same failure an operator gets by deploying before
    // `wrangler d1 migrations apply ferrogate-control`.
    await controlDb().prepare("DROP TABLE permissions").run();
    try {
      const res = await listPolicies();
      expect(res.status).toBe(503);
      expect(((await res.json()) as { error: { code: string } }).error.code).toBe(
        "rbac_unavailable",
      );
    } finally {
      await controlDb()
        .prepare(
          "CREATE TABLE IF NOT EXISTS permissions (id TEXT PRIMARY KEY, key TEXT NOT NULL UNIQUE, " +
            "name TEXT NOT NULL, description TEXT NOT NULL DEFAULT '', " +
            "created_at_unix INTEGER NOT NULL DEFAULT (unixepoch()), " +
            "updated_at_unix INTEGER NOT NULL DEFAULT (unixepoch()))",
        )
        .run();
    }
  });
});

/** Compile-time proof that a live `D1Database` satisfies the narrow port. */
const _bindingSatisfiesThePort: (db: D1Database) => RbacDatabase = (db) => db;
void _bindingSatisfiesThePort;
void (undefined as unknown as RbacDecision);
